src/LinuxTracing/PerfEventReaders.cpp

```diff
@@ -507,18 +510,22 @@ void ReadPerfSampleIdAll(PerfEventRingBuffer* ring_buffer, const perf_event_head
               .regs = std::move(res.regs),
               .dyn_size = res.dyn_size,
               .data = std::move(res.stack_data),
+              .just_tracepoint = !get_callstack,
```

Comment:
Oh I think I understand what's going on here. If the underlying PERF_RECORD_SAMPLE read from the ring buffer contains the stack, then you are producing a `SchedWakeupWithStackPerfEvent` no matter what, but then possibly not including the callstacks and setting `just_tracepoint`. This is conceptually wrong: if you don't get the callstack, a "regular" `SchedWakeupPerfEvent` should be produced (we don't need to keep "memory" of what the PERF_RECORD_SAMPLE actually contains). A separate reader function might be necessary.

Reply:
> A separate reader function might be necessary.

Alternatively `ConsumeSchedWakeupPerfEvent` could be able to read from a PERF_RECORD_SAMPLE whether it contains the stack and registers or not, possibly based on an extra flag parameter. Then you call `ConsumeSchedWakeupWithStackPerfEvent` or `ConsumeSchedWakeupPerfEvent` depending on `pid == target_pid_`. 

Does that make sense? What do you all think?